### ISSUE TITLE
[ProgressMonitor] Parent monitor/tasks do not track child progress

### ISSUE THREAD

**Issue by sevoku** [mono/monodevelop#7256]
When running long operations with an (initially) unknown amount of work, the progress of a child/sub- task is not propagated to the parent task and monitor, and the progress is either not shown or just stays at X% for a long time. Typical use cases:
* Downloads/Uploads
* Version Control remote operations

Given the following example from Git:
- VCS `Checkout` operation gets started (task: clone)
  - at this point VCS does not know the actual amount of work, that depends on the VCS backend implementation, so we start with 1 step
  - GIT Backend KNOWS how many steps it needs and starts a `Clone` task accordingly with 3 steps
    - Step 1: `Connect` and authenticate (no progress reported)
    - Step 2: `Fetch` from remote
      - only after connecting to the server we get the actual amount of objects to download and index
      - Step 2.1: `Download` objects (progress x objects downloaded)
      - Step 2.2: `Index objects` (progress x objects indexed)
    - Step 3: `Checkout` files (progress: x files checked out)

Problems in the current implementation:
 - after each Step in the Git backend we need to manually step the `Clone` task to bump the progress after each step ([failing test](https://github.com/mono/monodevelop/pull/7255/files#diff-71146016617dfa1ca78f4c78c801a7baR312))
  - it makes much more sense to update the progress automatically once a child task ends. This could be done either by assuming that a child task is exactly 1 step, or by adding an optional step parameter to `monitor.EndTask (int parentWorkDone = 1)`
- Progress change of a child is not picked up by a parent, `monitor.Step ()` with a child being the `monitor.CurrentTask` does not update `monitor.Progress` ([failing test](https://github.com/mono/monodevelop/pull/7255/files#diff-71146016617dfa1ca78f4c78c801a7baR347))

I have created #7255 based on these assumptions/expectations, since this looks like a bug to me, but this is up for discussion. At the end we need a reliable way to report a long running progress without actually knowing its amount of work before we start it.

**Comment by slluis** [comment]
So the way to report the progress for the checkout case could be something like this:
* VCS: 
    * BeginTask ("Clone", 1)
* Git:
   * BeginTask ("Git clone", 100) // Even though there are 3 steps, not all steps have the same duration. Let's consider that in general connection takes 10%, fetching 50% and checkout 40%.
       * ... Connect ...
        * Step (10) // Connection completed
        * BeginStep (50) // Start fetching
            * BeginTask (NUMBER_OF_OBJECTS)
                * Step * NUMBER_OF_OBJECTS
            * EndTask()
        * EndStep();
        * BeginStep (40) // Start checking out
            * BeginTask ("Checking out", NUMBER_OF_FILES)
                * Step * NUMBER_OF_FILES
           * EndTask ();
        * EndStep();
    * EndTask ();
* VCS:
   * EndTask ()

**Comment by sevoku** [comment]
No, this won't work, because the child tasks are not reported. No matter what we do within the Git impl, we have just one single task in VCS which is not aware of the actual work happening in the backend, so it stays at 0 for the whole process and jumpt to 100% at the very end.
We MIGHT initialize the root VCS task with `BeginTask ("Clone", 100)`, but then we still have the granularity of 3 tasks in the backend process (in your example it will start at 0 %, then jump to 10%, to 60% and 100%), it won't smoothly update the progress not even on 1% steps.
